storage/storage_dest.go

```diff
@@ -827,64 +790,128 @@ func (s *storageImageDestination) Commit(ctx context.Context, unparsedToplevel t
-			logrus.Debugf("error saving metadata for image %q: %v", img.ID, err)
-			return fmt.Errorf("saving metadata for image %q: %w", img.ID, err)
+		options.Metadata = string(metadata)
+	}
+
+	// Create the image record, pointing to the most-recently added layer.
+	intendedID := s.imageRef.id
+	if intendedID == "" {
+		intendedID = s.computeID(man)
+	}
+	oldNames := []string{}
+	img, err := s.imageRef.transport.store.CreateImage(intendedID, nil, lastLayer, "", options)
+	if err != nil {
+		if !errors.Is(err, storage.ErrDuplicateID) {
+			logrus.Debugf("error creating image: %q", err)
+			return fmt.Errorf("creating image %q: %w", intendedID, err)
 		}
-		logrus.Debugf("saved image metadata %q", string(metadata))
+		img, err = s.imageRef.transport.store.Image(intendedID)
+		if err != nil {
+			return fmt.Errorf("reading image %q: %w", intendedID, err)
+		}
+		if img.TopLayer != lastLayer {
+			logrus.Debugf("error creating image: image with ID %q exists, but uses different layers", intendedID)
+			return fmt.Errorf("image with ID %q already exists, but uses a different top layer: %w", intendedID, storage.ErrDuplicateID)
+		}
+		logrus.Debugf("reusing image ID %q", img.ID)
+		oldNames = append(oldNames, img.Names...)
+		// set the data items and metadata on the already-present image
+		for _, data := range options.BigData {
+			if err := s.imageRef.transport.store.SetImageBigData(img.ID, data.Key, data.Data, manifest.Digest); err != nil {
+				logrus.Debugf("error saving big data %q for image %q: %v", data.Key, img.ID, err)
+				return fmt.Errorf("saving big data %q for image %q: %w", data.Key, img.ID, err)
+			}
+		}
+		if options.Metadata != "" {
+			if err := s.imageRef.transport.store.SetMetadata(img.ID, options.Metadata); err != nil {
+				logrus.Debugf("error saving metadata for image %q: %v", img.ID, err)
+				return fmt.Errorf("saving metadata for image %q: %w", img.ID, err)
+			}
+			logrus.Debugf("saved image metadata %q", options.Metadata)
+		}
+	} else {
+		logrus.Debugf("created new image ID %q with metadata %q", img.ID, options.Metadata)
 	}
-	// Adds the reference's name on the image.  We don't need to worry about avoiding duplicate
+
+	// Clean up the unfinished image on any error.
+	// (Is this the right thing to do if the image has existed before?)
+	commitSucceeded := false
+	defer func() {
+		if !commitSucceeded {
+			logrus.Errorf("Updating image %q (old names %v) failed, deleting it", img.ID, oldNames)
+			if _, err := s.imageRef.transport.store.DeleteImage(img.ID, true); err != nil {
+				logrus.Errorf("Error deleting incomplete image %q: %v", img.ID, err)
+			}
+		}
+	}()
+
+	// Add the reference's name on the image.  We don't need to worry about avoiding duplicate
 	// values because AddNames() will deduplicate the list that we pass to it.
 	if name := s.imageRef.DockerReference(); name != nil {
 		if err := s.imageRef.transport.store.AddNames(img.ID, []string{name.String()}); err != nil {
```

Comment:
This can also be done using a parameter to `CreateImage` if the image does not exist yet; that would avoid another trip through the fairly costly locking implementation.

---

… and then we would really need the `commitSucceeded` handling _only_ in the “image already exists” case, and that would be an opportunity revisit whether we really want to delete the image, or maybe we can just delete that code.

Reply:
Some care has to be taken with that parameter - if any of the names passed at create-time are already in use, we get an `ErrDuplicateName` error back, and I think we want to avoid that.

Reply:
Oops, I didn’t notice that `CreateImage` returns `ErrDuplicateName`.

It might make sense to let callers opt in into overriding names with a new field in `storage.ImageOptions`, but that would not be quite an one-line change, so let’s not wait on that.